storage/src/main/java/org/apache/kafka/storage/internals/log/ProducerStateManager.java

```diff
@@ -0,0 +1,676 @@
+                long currentTxnFirstOffset = producerEntryStruct.getLong(CURRENT_TXN_FIRST_OFFSET_FIELD);
+
+                OptionalLong currentTxnFirstOffsetVal = currentTxnFirstOffset >= 0 ? OptionalLong.of(currentTxnFirstOffset) : OptionalLong.empty();
+                Optional<BatchMetadata> batchMetadata =
+                        (offset >= 0) ? Optional.of(new BatchMetadata(seq, offset, offsetDelta, timestamp)) : Optional.empty();
+                entries.add(new ProducerStateEntry(producerId, producerEpoch, coordinatorEpoch, timestamp, currentTxnFirstOffsetVal, batchMetadata));
+            }
+
+            return entries;
+        } catch (SchemaException e) {
+            throw new CorruptSnapshotException("Snapshot failed schema validation: " + e.getMessage());
+        }
+    }
+
+    private static void writeSnapshot(File file, Map<Long, ProducerStateEntry> entries) throws IOException {
+        Struct struct = new Struct(PID_SNAPSHOT_MAP_SCHEMA);
+        struct.set(VERSION_FIELD, PRODUCER_SNAPSHOT_VERSION);
+        struct.set(CRC_FIELD, 0L); // we'll fill this after writing the entries
+        Struct[] structEntries = new Struct[entries.size()];
+        int i = 0;
+        for (Map.Entry<Long, ProducerStateEntry> producerIdEntry : entries.entrySet()) {
+            Long producerId = producerIdEntry.getKey();
+            ProducerStateEntry entry = producerIdEntry.getValue();
+            Struct producerEntryStruct = struct.instance(PRODUCER_ENTRIES_FIELD);
+            producerEntryStruct.set(PRODUCER_ID_FIELD, producerId)
+                    .set(PRODUCER_EPOCH_FIELD, entry.producerEpoch())
+                    .set(LAST_SEQUENCE_FIELD, entry.lastSeq())
+                    .set(LAST_OFFSET_FIELD, entry.lastDataOffset())
+                    .set(OFFSET_DELTA_FIELD, entry.lastOffsetDelta())
+                    .set(TIMESTAMP_FIELD, entry.lastTimestamp())
+                    .set(COORDINATOR_EPOCH_FIELD, entry.coordinatorEpoch())
+                    .set(CURRENT_TXN_FIRST_OFFSET_FIELD, entry.currentTxnFirstOffset().orElse(-1L));
+            structEntries[i++] = producerEntryStruct;
+        }
+        struct.set(PRODUCER_ENTRIES_FIELD, structEntries);
+
+        ByteBuffer buffer = ByteBuffer.allocate(struct.sizeOf());
+        struct.writeTo(buffer);
+        buffer.flip();
+
+        // now fill in the CRC
+        long crc = Crc32C.compute(buffer, PRODUCER_ENTRIES_OFFSET, buffer.limit() - PRODUCER_ENTRIES_OFFSET);
+        ByteUtils.writeUnsignedInt(buffer, CRC_OFFSET, crc);
+
+        try (FileChannel fileChannel = FileChannel.open(file.toPath(), StandardOpenOption.CREATE, StandardOpenOption.WRITE)) {
+            fileChannel.write(buffer);
+            fileChannel.force(true);
+        }
+    }
+
+    private static boolean isSnapshotFile(Path path) {
+        File file = path.toFile();
+        return file.isFile() && file.getName().endsWith(LogFileUtils.PRODUCER_SNAPSHOT_FILE_SUFFIX);
```

Comment:
I think we can implement this directly on the path without converting to File.

Reply:
I am not sure there is a way to find whether a given path instance is a `file` or not like the way we check `File.isFile()`.

Reply:
How about `Files.isRegularFile(path)`?

Reply:
Thanks for that. I will make the change and update the PR.